python-package/lightgbm/sklearn.py

```diff
@@ -151,17 +181,25 @@ def __call__(self, preds: np.ndarray, dataset: Dataset) -> Tuple[np.ndarray, np.
             The value of the second order derivative (Hessian) of the loss
             with respect to the elements of preds for each sample point.
         """
-        labels = dataset.get_label()
+        labels = _get_label_from_constructed_dataset(dataset)
         argc = len(signature(self.func).parameters)
         if argc == 2:
             grad, hess = self.func(labels, preds)  # type: ignore[call-arg]
-        elif argc == 3:
-            grad, hess = self.func(labels, preds, dataset.get_weight())  # type: ignore[call-arg]
-        elif argc == 4:
-            grad, hess = self.func(labels, preds, dataset.get_weight(), dataset.get_group())  # type: ignore [call-arg]
-        else:
-            raise TypeError(f"Self-defined objective function should have 2, 3 or 4 arguments, got {argc}")
-        return grad, hess
+            return grad, hess
+
+        weight = _get_weight_from_constructed_dataset(dataset)
+        if argc == 3:
+            grad, hess = self.func(labels, preds, weight)  # type: ignore[call-arg]
+            return grad, hess
+
+        if argc == 4:
+            group = _get_group_from_constructed_dataset(dataset)
+            if group is not None:
+                return self.func(labels, preds, weight, group)  # type: ignore[call-arg]
+            else:
+                return self.func(labels, preds, weight, group)  # type: ignore[call-arg]
```

Comment:
I think these two are identical now
```suggestion
            return self.func(labels, preds, weight, group)  # type: ignore[call-arg]
```

Reply:
😂 swear this was just a result of not being careful after a refactoring, not just me trying to cheat one how many lines of code I've committed haha